iron-fit-behavior.html

```diff
@@ -195,38 +301,123 @@
+        var areaOverlapTop = this.__getCroppedArea({
+          top: top,
+          left: noOverlapLeft,
+          width: rect.width,
+          height: rect.height
+        }, fitRect);
+        if (areaOverlapLeft < areaOverlapTop) {
+          left = noOverlapLeft;
+        } else {
+          top = noOverlapTop;
+        }
+      }
+
+      var right = left + rect.width;
+      var bottom = top + rect.height;
+
+      left = Math.max(left, 0);
+      top = Math.max(top, 0);
+
+      var maxWidth = Math.min(fitRect.right, right) - left;
+      var maxHeight = Math.min(fitRect.bottom, bottom) - top;
+
+      var minWidth = this._fitInfo.sizedBy.minWidth;
+      var minHeight = this._fitInfo.sizedBy.minHeight;
+
+      if (maxWidth < minWidth || right < minWidth) {
+        left = -9999;
+        maxWidth = 0;
+      }
+      if (maxHeight < minHeight || bottom < minHeight) {
+        top = -9999;
+        maxHeight = 0;
+      }
+
+      this.sizingTarget.style.maxWidth = maxWidth + 'px';
+      this.sizingTarget.style.maxHeight = maxHeight + 'px';
+
+      // Remove the offset caused by any stacking context.
+      this.style.left = (left - rect.left) + 'px';
+      this.style.top = (top - rect.top) + 'px';
+    },
+
+    /**
      * Constrains the size of the element to the window or `fitInfo` by setting `max-height`
      * and/or `max-width`.
      */
     constrain: function() {
+      if (this.horizontalAlign || this.verticalAlign) {
+        return;
+      }
       var info = this._fitInfo;
       // position at (0px, 0px) if not already positioned, so we can measure the natural size.
-      if (!this._fitInfo.positionedBy.vertically) {
+      if (!info.positionedBy.vertically) {
+        this.style.position = 'fixed';
         this.style.top = '0px';
       }
-      if (!this._fitInfo.positionedBy.horizontally) {
-        this.style.left = '0px';
-      }
-      if (!this._fitInfo.positionedBy.vertically || !this._fitInfo.positionedBy.horizontally) {
-        // need position:fixed to properly size the element
+      if (!info.positionedBy.horizontally) {
         this.style.position = 'fixed';
+        this.style.left = '0px';
       }
+
       // need border-box for margin/padding
       this.sizingTarget.style.boxSizing = 'border-box';
       // constrain the width and height if not already set
       var rect = this.getBoundingClientRect();
+      var fitRect = this.__getNormalizedRect(this.fitInto);
       if (!info.sizedBy.height) {
-        this._sizeDimension(rect, info.positionedBy.vertically, 'top', 'bottom', 'Height');
+        this._sizeDimension(rect, fitRect, 'Height');
       }
       if (!info.sizedBy.width) {
-        this._sizeDimension(rect, info.positionedBy.horizontally, 'left', 'right', 'Width');
+        this._sizeDimension(rect, fitRect, 'Width');
       }
     },
 
-    _sizeDimension: function(rect, positionedBy, start, end, extent) {
+    _sizeDimension: function(rect, fitRect, extent) {
```

Comment:
I think that we should probably mark this method as private. Right now it's protected, so technically this would be a breaking change.